File: point_controller/point_controller/point_controller_node.py
"""ROS 2 node that drives the robot toward a single map-frame goal pose."""

from __future__ import annotations

from math import atan2

from geometry_msgs.msg import PoseStamped, Twist
import rclpy
from rclpy.node import Node
from rclpy.time import Time
from std_msgs.msg import String
from tf2_ros import Buffer, TransformException, TransformListener

from point_controller.control import ControllerConfig, GoalPose, RobotPose, compute_command


def quaternion_to_yaw(q) -> float:
    return atan2(
        2.0 * (q.w * q.z + q.x * q.y),
        1.0 - 2.0 * (q.y * q.y + q.z * q.z),
    )


class PointControllerNode(Node):
    def __init__(self) -> None:
        super().__init__('point_controller_node')

        self.declare_parameter('global_frame_id', 'map')
        self.declare_parameter('base_frame_id', 'base_link')
        self.declare_parameter('control_rate_hz', 20.0)
        self.declare_parameter('xy_tolerance_m', 0.05)
        self.declare_parameter('yaw_tolerance_rad', 0.0872665)
        self.declare_parameter('heading_tolerance_rad', 0.15)
        self.declare_parameter('k_v', 0.8)
        self.declare_parameter('k_w', 1.8)
        self.declare_parameter('v_max_m_s', 0.25)
        self.declare_parameter('w_max_rad_s', 0.8)
        self.declare_parameter('v_min_m_s', 0.03)
        self.declare_parameter('w_min_rad_s', 0.08)

        self.global_frame_id = self.get_parameter('global_frame_id').value
        self.base_frame_id = self.get_parameter('base_frame_id').value
        self.config = ControllerConfig(
            xy_tolerance_m=float(self.get_parameter('xy_tolerance_m').value),
            yaw_tolerance_rad=float(self.get_parameter('yaw_tolerance_rad').value),
            heading_tolerance_rad=float(self.get_parameter('heading_tolerance_rad').value),
            k_v=float(self.get_parameter('k_v').value),
            k_w=float(self.get_parameter('k_w').value),
            v_max_m_s=float(self.get_parameter('v_max_m_s').value),
            w_max_rad_s=float(self.get_parameter('w_max_rad_s').value),
            v_min_m_s=float(self.get_parameter('v_min_m_s').value),
            w_min_rad_s=float(self.get_parameter('w_min_rad_s').value),
        )

        self.tf_buffer = Buffer()
        self.tf_listener = TransformListener(self.tf_buffer, self)
        self.goal = None
        self.last_phase = ''

        self.goal_sub = self.create_subscription(PoseStamped, '/goal_pose', self.on_goal_pose, 10)
        self.cmd_pub = self.create_publisher(Twist, '/cmd_vel', 10)
        self.status_pub = self.create_publisher(String, '/point_controller/status', 10)

        control_rate_hz = float(self.get_parameter('control_rate_hz').value)
        self.create_timer(max(0.01, 1.0 / control_rate_hz), self.control_step)

    def on_goal_pose(self, msg: PoseStamped) -> None:
        frame_id = msg.header.frame_id or self.global_frame_id
        if frame_id != self.global_frame_id:
            self.get_logger().warn(
                f'Ignoring goal in frame "{frame_id}"; expected "{self.global_frame_id}"'
            )
            return

        self.goal = GoalPose(
            x=msg.pose.position.x,
            y=msg.pose.position.y,
            yaw=quaternion_to_yaw(msg.pose.orientation),
        )
        self.last_phase = ''
        self.get_logger().info(
            f'Accepted goal x={self.goal.x:.3f}, y={self.goal.y:.3f}, yaw={self.goal.yaw:.3f}'
        )

    def control_step(self) -> None:
        if self.goal is None:
            return

        try:
            transform = self.tf_buffer.lookup_transform(
                self.global_frame_id,
                self.base_frame_id,
                Time(),
            )
        except TransformException as exc:
            self.get_logger().warn(
                f'Cannot lookup {self.global_frame_id}->{self.base_frame_id}: {exc}',
                throttle_duration_sec=1.0,
            )
            self.publish_stop()
            return

        robot = RobotPose(
            x=transform.transform.translation.x,
            y=transform.transform.translation.y,
            yaw=quaternion_to_yaw(transform.transform.rotation),
        )
        result = compute_command(robot, self.goal, self.config)

        if result.phase != self.last_phase:
            self.last_phase = result.phase
            self.get_logger().info(
                f'Point controller phase={result.phase}, '
                f'distance={result.distance_error:.3f}, yaw_error={result.yaw_error:.3f}'
            )

        twist = Twist()
        twist.linear.x = result.v
        twist.angular.z = result.w
        self.cmd_pub.publish(twist)

        status = String()
        status.data = (
            f'phase={result.phase} distance={result.distance_error:.3f} '
            f'yaw_error={result.yaw_error:.3f} v={result.v:.3f} w={result.w:.3f}'
        )
        self.status_pub.publish(status)

        if result.reached:
            self.goal = None

    def publish_stop(self) -> None:
        self.cmd_pub.publish(Twist())


def main(args=None) -> None:
    rclpy.init(args=args)
    node = PointControllerNode()
    try:
        rclpy.spin(node)
    finally:
        node.publish_stop()
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: point_controller/src/point_controller_node.cpp
#include "point_controller/control.hpp"

#include <geometry_msgs/msg/pose_stamped.hpp>
#include <geometry_msgs/msg/twist.hpp>
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>
#include <tf2/time.hpp>
#include <tf2_ros/buffer.h>
#include <tf2_ros/transform_listener.h>

#include <chrono>
#include <cmath>
#include <memory>
#include <optional>
#include <sstream>
#include <string>

namespace
{

double quaternion_to_yaw(const geometry_msgs::msg::Quaternion & q)
{
  return std::atan2(
    2.0 * (q.w * q.z + q.x * q.y),
    1.0 - 2.0 * (q.y * q.y + q.z * q.z));
}

}  // namespace

class PointControllerNode : public rclcpp::Node
{
public:
  PointControllerNode()
  : Node("point_controller_node"),
    tf_buffer_(get_clock()),
    tf_listener_(tf_buffer_)
  {
    global_frame_id_ = declare_parameter<std::string>("global_frame_id", "car_carto_map");
    base_frame_id_ = declare_parameter<std::string>("base_frame_id", "car_base_link");
    const double control_rate_hz = declare_parameter<double>("control_rate_hz", 20.0);
    config_.xy_tolerance_m = declare_parameter<double>("xy_tolerance_m", 0.05);
    config_.yaw_tolerance_rad = declare_parameter<double>("yaw_tolerance_rad", 0.0872665);
    config_.heading_tolerance_rad =
      declare_parameter<double>("heading_tolerance_rad", 0.15);
    config_.k_v = declare_parameter<double>("k_v", 0.8);
    config_.k_w = declare_parameter<double>("k_w", 1.8);
    config_.v_max_m_s = declare_parameter<double>("v_max_m_s", 0.25);
    config_.w_max_rad_s = declare_parameter<double>("w_max_rad_s", 0.8);
    config_.v_min_m_s = declare_parameter<double>("v_min_m_s", 0.03);
    config_.w_min_rad_s = declare_parameter<double>("w_min_rad_s", 0.08);

    goal_sub_ = create_subscription<geometry_msgs::msg::PoseStamped>(
      "/goal_pose", 10,
      std::bind(&PointControllerNode::on_goal_pose, this, std::placeholders::_1));
    cmd_pub_ = create_publisher<geometry_msgs::msg::Twist>("/cmd_vel", 10);
    status_pub_ = create_publisher<std_msgs::msg::String>("/point_controller/status", 10);

    const double period_s = std::max(0.01, 1.0 / control_rate_hz);
    timer_ = create_wall_timer(
      std::chrono::duration_cast<std::chrono::nanoseconds>(
        std::chrono::duration<double>(period_s)),
      std::bind(&PointControllerNode::control_step, this));
  }

  ~PointControllerNode() override
  {
    publish_stop();
  }

private:
  void on_goal_pose(const geometry_msgs::msg::PoseStamped::SharedPtr msg)
  {
    const auto frame_id =
      msg->header.frame_id.empty() ? global_frame_id_ : msg->header.frame_id;
    if (frame_id != global_frame_id_) {
      RCLCPP_WARN(
        get_logger(), "Ignoring goal in frame \"%s\"; expected \"%s\"",
        frame_id.c_str(), global_frame_id_.c_str());
      return;
    }

    goal_ = point_controller::GoalPose{
      msg->pose.position.x,
      msg->pose.position.y,
      quaternion_to_yaw(msg->pose.orientation)};
    last_phase_.clear();
    RCLCPP_INFO(
      get_logger(), "Accepted goal x=%.3f, y=%.3f, yaw=%.3f",
      goal_->x, goal_->y, goal_->yaw);
  }

  void control_step()
  {
    if (!goal_) {
      return;
    }

    geometry_msgs::msg::TransformStamped transform;
    try {
      transform = tf_buffer_.lookupTransform(
        global_frame_id_, base_frame_id_, tf2::TimePointZero);
    } catch (const tf2::TransformException & exc) {
      RCLCPP_WARN_THROTTLE(
        get_logger(), *get_clock(), 1000, "Cannot lookup %s->%s: %s",
        global_frame_id_.c_str(), base_frame_id_.c_str(), exc.what());
      publish_stop();
      return;
    }

    const point_controller::RobotPose robot{
      transform.transform.translation.x,
      transform.transform.translation.y,
      quaternion_to_yaw(transform.transform.rotation)};
    const auto result = point_controller::compute_command(robot, *goal_, config_);

    if (result.phase != last_phase_) {
      last_phase_ = result.phase;
      RCLCPP_INFO(
        get_logger(), "Point controller phase=%s, distance=%.3f, yaw_error=%.3f",
        result.phase.c_str(), result.distance_error, result.yaw_error);
    }

    geometry_msgs::msg::Twist twist;
    twist.linear.x = result.v;
    twist.angular.z = result.w;
    cmd_pub_->publish(twist);

    std::ostringstream status;
    status << "phase=" << result.phase
           << " distance=" << result.distance_error
           << " yaw_error=" << result.yaw_error
           << " v=" << result.v
           << " w=" << result.w;
    std_msgs::msg::String status_msg;
    status_msg.data = status.str();
    status_pub_->publish(status_msg);

    if (result.reached) {
      goal_.reset();
    }
  }

  void publish_stop()
  {
    if (cmd_pub_) {
      cmd_pub_->publish(geometry_msgs::msg::Twist{});
    }
  }

  std::string global_frame_id_;
  std::string base_frame_id_;
  point_controller::ControllerConfig config_;
  std::optional<point_controller::GoalPose> goal_;
  std::string last_phase_;

  tf2_ros::Buffer tf_buffer_;
  tf2_ros::TransformListener tf_listener_;
  rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr goal_sub_;
  rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr cmd_pub_;
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr status_pub_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<PointControllerNode>());
  rclcpp::shutdown();
  return 0;
}
